crates/edr_solidity/src/bytecode_trie.rs

```diff
@@ -1,83 +1,417 @@
-        // of them. Note that this includes the metadata hash, so the chances of
-        // happening are pretty remote, except in super artificial cases that we
-        // have in our test suite.
-        cursor.match_ = Some(bytecode.clone());
+        // If multiple contracts with the exact same bytecode are added we keep
+        // the last of them. Note that this includes the metadata hash,
+        // so the chances of happening are pretty remote, except in
+        // super artificial cases that we have in our test suite.
+        cursor.match_ = Some(new_item);
     }
 
     /// Searches for a bytecode. If it's an exact match, it is returned. If
     /// there's no match, but a prefix of the code is found in the trie, the
     /// node of the longest prefix is returned. If the entire code is
     /// covered by the trie, and there's no match, we return None.
-    pub fn search(&self, code: &[u8], current_code_byte: u32) -> Option<TrieSearch<'_>> {
-        if current_code_byte > code.len() as u32 {
+    pub fn search(&self, key: &[u8], current_code_byte: usize) -> Option<TrieSearch<'_, T>> {
+        if current_code_byte > key.len() {
             return None;
         }
 
         let mut cursor = self;
+        let mut index = current_code_byte;
 
-        for byte in code.iter().skip(current_code_byte as usize) {
-            let child_node = cursor.child_nodes.get(byte);
-
-            if let Some(node) = child_node {
+        while index < key.len() {
+            if index < cursor.prefix.range_end {
+                if key[index] != cursor.prefix.key.key()[index] {
+                    // Cursor cannot be root here, because the root's prefix ends at index 0.
+                    return Some(TrieSearch::LongestPrefixNode {
+                        node: cursor,
+                        diff_index: index,
+                        // We're not yet at the end of the range, so it cannot be a match.
+                        match_: None,
+                    });
+                }
+            } else if let Some(node) = cursor.child_nodes.get(&key[index]) {
                 cursor = node;
+            } else if !cursor.is_root() {
+                return Some(TrieSearch::LongestPrefixNode {
+                    node: cursor,
+                    diff_index: index,
+                    match_: cursor.match_.clone(),
+                });
+            }
+
+            index += 1;
+        }
+
+        cursor.match_.as_ref().and_then(|item| {
+            if cursor.prefix.range_end == key.len() {
+                // If the cursor's range ends where the key ends, we have a hit.
+                Some(TrieSearch::ExactHit(item.clone()))
             } else {
-                return Some(TrieSearch::LongestPrefixNode(cursor));
+                // Otherwise the cursor's range is greater than the key's length which means the
+                // key is a prefix of the match.
+                None
```

Comment:
Is this the case where the key is "foo" and a stored entry is "foobar"?

Wouldn't we still be interested in that case? Or is there a separate function for that?

Reply:
> Is this the case where the key is "foo" and a stored entry is "foobar"?
Yes, we'd return `None` in this case.

> Wouldn't we still be interested in that case? Or is there a separate function for that?
Yes, normally this is an interesting result for trie lookups, but for the purposes of the contract identifier, it's a no match, so we're following the previous implementation here by returning `None`.